### PR TITLE
Added python decoder

### PR DESCRIPTION
**Pull request by Narasimha1997**
@lautat 
This PR adds `decode.c` and Python wrapper for the same. 
The following functions are added newly :
`decode` : Which takes blurhash, width, height, punch and mode as input and returns `PIL.Image` object which can be used to manipulate the image further by the caller. Here `mode` parameter is either `RGB` or `RGBA`. (3 or 4 channels)

`is_valid_blurhash` : Validates the blurhash, this function can be used by those who just want to know if they are dealing with the correct blurhash string, without decoding.

Let me know if you need any changes.

Thanks and Regards,
Prasanna

### CODE REVIEW

**Comment by iomintz** [comment]
Again the wrapper library should use python's memory allocation system rather than libc malloc() / free().

**Comment by Narasimha1997** [comment]
@iomintz , yes you are right.
Please check the commit 
3dc8265

File: src/build_blurhash.py
Comment (lautat) [inline src/build_blurhash.py:8]: `_functions` might be a bit too generic name, but I'll have to think a bit to come up with a good alternative. It is fine for now since renaming it later is not an issue.

File: src/blurhash/__init__.py
Comment (lautat) [inline src/blurhash/__init__.py:67]: I would use a subclass of `Exception` in this case, for example, `BlurhashDecodeError`. Then, it is easier to handle all possible errors with separate `expect` statements without catching other exceptions that might be unexpected.

File: src/blurhash/__init__.py
Comment (lautat) [inline src/blurhash/__init__.py:74]: Prefix the name of this function with an underscore since it seems that this is not meant to be part of the public interface of this library. If you think that this should be part of it, then it makes sense to include it in  `__all__`.

**Review by lautat — CHANGES_REQUESTED** [review]
Thank you for this contribution! Based on a brief look, the Python part looks great. I've made a few suggestions on how the implementation could be improved. In addition, please check the code style with `flake8` fix if necessary to be sure that the new code is consistent with the existing code. Using it with default settings is fine. (I should configure it and document the usage in `README.md`.) The `decode` function should also have a few tests. It probably does not make sense to verify that the result is expected by a `==` check as in the `encode` test cases, but at least the tests can verify parameter validation and dimensions of the returned image object. That should help ensure that the `decode` function works even when the implementation is changed. Finally, it would be nice to have some usage examples of the `decode` function in `README.md` so that it is clear for the library user what kind of parameters are expected and what the function returns. I'll likely take another proper look at the code, and especially at the C FFI, a bit later in case I find something else that could be improved.

I will wait for the approval of woltapp/blurhash#94 before merging this. Unfortunately, it may take a while. I'll likely have time to have a look at it soon, but I am not familiar with the decoding algorithm so I've tried to find someone else to check it.

File: src/blurhash/__init__.py
Comment (Narasimha1997) [reply src/blurhash/__init__.py:74]: @lautat , yes this is supposed to be a public-function, added it in `__all__`, this function allows anyone to check if the blurhash they have is valid or not without calling `decode` function.

File: src/build_blurhash.py
Comment (Narasimha1997) [reply src/build_blurhash.py:8]: Yeah! I couldn't come up with a proper name for this, so leaving this as `_functions` as of now.

File: src/blurhash/__init__.py
Comment (Narasimha1997) [reply src/blurhash/__init__.py:48]: @lautat 
Isn't `Enum` is going to throw `KeyError` if the given symbol is not found in the enumeration??

File: src/blurhash/__init__.py
Comment (Narasimha1997) [reply src/blurhash/__init__.py:67]: Yeah added `BlurhashDecodeError` sub-class to resolve this.

File: src/blurhash/__init__.py
Comment (Narasimha1997) [reply src/blurhash/__init__.py]: Handled this as well.

File: src/blurhash/__init__.py
Comment (Narasimha1997) [reply src/blurhash/__init__.py]: `RGBA` mode just adds another channel with default value `0xff`, some of the use-cases require `RGBA` image so that it would be convenient to some further image processing operations from the pixels obtained, if this is not there, we have to manually apply transforms on `PIL.Image.Image` again, so it would be a helpful feature I thought. (This is just a step during decoding and does not depend on encoded blurhash)

**Comment by Narasimha1997** [comment]
@lautat 
Thanks for suggesting the changes, I have incorporated them in the latest commit c1452bf

And btw, the C decoder algorithm is written keeping TypeScript implementation of the decoder as a reference.
Maybe you can just compare the steps in both C and TypeScript to check the algorithm's correctness.

Regards,
Prasanna

**Comment by lautat** [comment]
> And btw, the C decoder algorithm is written keeping TypeScript implementation of the decoder as a reference.
Maybe you can just compare the steps in both C and TypeScript to check the algorithm's correctness.

Thanks! I'll keep this in mind when I'm checking it.

**Comment by Narasimha1997** [comment]
@lautat , 
Any updates on this??

**Comment by lautat** [comment]
This looks good, but one more thing to fix: check the output of flake8, and please try to fix the errors it reports. That helps to keep the code style consistent. On `master` branch it outputs one warning, but here it gives more. There are tools like `black` which can do the formatting automatically, but I would not use them in this case since the existing code hasn't been automatically formatted so these tools likely do more changes than necessary. Let me know if you need help with understanding the output or fixing errors. I'll try to check woltapp/blurhash#94 soon-ish, but probably not before Friday.
```
./src/build_blurhash.py:10:80: E501 line too long (92 > 79 characters)
./src/build_blurhash.py:13:80: E501 line too long (93 > 79 characters)
./src/build_blurhash.py:24:80: E501 line too long (81 > 79 characters)
./src/build_blurhash.py:25:80: E501 line too long (87 > 79 characters)
./src/build_blurhash.py:26:80: E501 line too long (90 > 79 characters)
./src/build_blurhash.py:32:80: E501 line too long (116 > 79 characters)
./src/build_blurhash.py:38:80: E501 line too long (81 > 79 characters)
./src/build_blurhash.py:39:80: E501 line too long (85 > 79 characters)
./src/blurhash/__init__.py:14:80: E501 line too long (100 > 79 characters)
./src/blurhash/__init__.py:16:1: E302 expected 2 blank lines, found 1
./src/blurhash/__init__.py:20:1: E302 expected 2 blank lines, found 1
./src/blurhash/__init__.py:20:37: E203 whitespace before ':'
./src/blurhash/__init__.py:53:1: E302 expected 2 blank lines, found 1
./src/blurhash/__init__.py:53:42: E251 unexpected spaces around keyword / parameter equals
./src/blurhash/__init__.py:53:44: E251 unexpected spaces around keyword / parameter equals
./src/blurhash/__init__.py:53:52: E251 unexpected spaces around keyword / parameter equals
./src/blurhash/__init__.py:53:54: E251 unexpected spaces around keyword / parameter equals
./src/blurhash/__init__.py:56:80: E501 line too long (106 > 79 characters)
./src/blurhash/__init__.py:59:80: E501 line too long (108 > 79 characters)
./src/blurhash/__init__.py:62:80: E501 line too long (107 > 79 characters)
./src/blurhash/__init__.py:65:80: E501 line too long (103 > 79 characters)
./src/blurhash/__init__.py:70:48: E203 whitespace before ':'
./src/blurhash/__init__.py:76:5: F841 local variable 'channels_int' is assigned to but never used
./src/blurhash/__init__.py:80:80: E501 line too long (82 > 79 characters)
./src/blurhash/__init__.py:81:37: E128 continuation line under-indented for visual indent
./src/blurhash/__init__.py:83:20: E203 whitespace before ':'
./src/blurhash/__init__.py:91:1: E302 expected 2 blank lines, found 1
./src/blurhash/__init__.py:91:32: E203 whitespace before ':'
./tests/test_decode.py:8:1: E302 expected 2 blank lines, found 1
./tests/test_decode.py:14:66: E251 unexpected spaces around keyword / parameter equals
./tests/test_decode.py:14:68: E251 unexpected spaces around keyword / parameter equals
./tests/test_decode.py:14:80: E501 line too long (84 > 79 characters)
./tests/test_decode.py:14:83: E202 whitespace before ')'
./tests/test_decode.py:17:1: E302 expected 2 blank lines, found 1
./tests/test_decode.py:18:66: E251 unexpected spaces around keyword / parameter equals
./tests/test_decode.py:18:68: E251 unexpected spaces around keyword / parameter equals
./tests/test_decode.py:18:80: E501 line too long (82 > 79 characters)
./tests/test_decode.py:21:1: E302 expected 2 blank lines, found 1
./tests/test_decode.py:22:67: E251 unexpected spaces around keyword / parameter equals
./tests/test_decode.py:22:69: E251 unexpected spaces around keyword / parameter equals
./tests/test_decode.py:25:1: E302 expected 2 blank lines, found 1
./tests/test_decode.py:29:1: E302 expected 2 blank lines, found 1
./tests/test_decode.py:31:62: E251 unexpected spaces around keyword / parameter equals
./tests/test_decode.py:31:64: E251 unexpected spaces around keyword / parameter equals
./tests/test_decode.py:33:1: E302 expected 2 blank lines, found 1
./tests/test_decode.py:42:1: E302 expected 2 blank lines, found 1
./tests/test_decode.py:44:63: E251 unexpected spaces around keyword / parameter equals
./tests/test_decode.py:44:65: E251 unexpected spaces around keyword / parameter equals
```

**Comment by Narasimha1997** [comment]
@lautat 
Cool thanks, please take your time. 😊

**Comment by Narasimha1997** [comment]
@lautat 
In the recent commit, I have fixed `flake8` convention inconsistencies, I have also verified the same using `flake8` tool.

**Review by lautat — APPROVED** [review]
Thank you very much for this contribution! I will review the C implementation separately in the pull request you opened in the blurhash repository. I expect that you update the C implementation included here once the review is done, but I don't think there will be any major changes affecting the Python interface. We'll continue the discussion in the other pull request once I have a chance to review it. When all that is done, I will merge this pull request and create a release.

**Comment by Narasimha1997** [comment]
@lautat , Thanks.

**Comment by Narasimha1997** [comment]
@lautat 
Hey, Since the PR to the original blurhash repository got approved and merged. I have made similar changes to this repository as well. I have verified `flake8` conventions and ran unit-tests to make sure things didn't break after modifications.
Thanks.

**Review by lautat — APPROVED** [review]
Thank you very much for your contribution!

**Comment by lautat** [comment]
@Narasimha1997, I have now released a new version of the library including these changes: https://pypi.org/project/blurhash-python/1.1.0/.

**Comment by Narasimha1997** [comment]
@lautat 
Cool, thank you 👍

**Pull request merged**
